### PR TITLE
PLAT-1044-3 Expose the DomPopup hierarchy via IDomPopupCompositor

### PR DESCRIPTION
**Pull request by alexander-schmidt**

### CODE REVIEW

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java [hunk 1]
```diff
@@ -67,13 +42,27 @@ public List<DomPopup> getAllChildPopups(DomPopup parent) {
...
-	 * Removes all {@link DomPopup} elements from this
-	 * {@link DomPopupHierarchyTree}.
+	 * Adds the given parent-child relation to this tree.
+	 *
+	 * @param parent
+	 *            the parent {@link DomPopup} (never <i>null</i>)
+	 * @param child
+	 *            the child {@link DomPopup} (never <i>null</i>)
+	 * @throws IllegalArgumentException
+	 *             if the given parent and child references are the same
```
Comment (oliver-richers) [inline platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java:60]: That's a lie! At least, I cannot see, where this is thrown. ;)

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java (hunk 1 above)
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java:60]:
```
private static class ChildMap extends HashMap<DomPopup, List<DomPopup>> {

public void put(DomPopup parent, DomPopup child) {

	Objects.requireNonNull(parent);
	Objects.requireNonNull(child);
	if (parent == child) {
		throw new IllegalArgumentException();           <---------- here
	}
	computeIfAbsent(parent, dummy -> new ArrayList<>()).add(child);
}
}
```

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.softicar.platform.dom.elements.popup.compositor;
+
+import com.softicar.platform.dom.elements.popup.DomPopup;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A tree that contains parent-child nesting relations between {@link DomPopup}
+ * elements.
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java:9]: It does not. Removed.

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	 *
+	 * @param parent
+	 *            the parent {@link DomPopup} (never <i>null</i>)
+	 * @return all child {@link DomPopup} elements (never <i>null</i>)
+	 */
+	List<DomPopup> getAllChildPopups(DomPopup parent);
+
+	/**
+	 * Returns the {@link DomPopup} from which the given {@link DomPopup} was
+	 * spawned.
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java:34]: Actually, the spawning order is insufficient to explain this. Expanded the Javadoc on top of `IDomPopupHierarchyTree`, pls check it out.

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public interface IDomPopupHierarchyTree {
+
+	/**
+	 * Determines all transitive child {@link DomPopup} elements of the given
+	 * parent {@link DomPopup}.
+	 * <p>
+	 * The returned {@link List} will be ordered according to descending depth
+	 * in the hierarchy tree. That is, deeply-nested children will occur before
+	 * less deeply-nested children.
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/IDomPopupHierarchyTree.java:21]: reworded the javadoc accordingly

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java [hunk 5]
```diff
@@ -3,59 +3,34 @@
...
+import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 /**
  * A tree that contains parent-child nesting relations between {@link DomPopup}
  * elements.
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/compositor/DomPopupHierarchyTree.java:15]: ha, indeed

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**